[Hexagon][llvm-objdump] Improve disassembly of Hexagon bundles

quic-areg opened the pull request:
Hexagon instructions are VLIW "bundles" of up to four instruction words encoded as a single MCInst with operands for each sub-instruction. Previously, the disassembler's getInstruction() returned the full bundle, which made it difficult to work with llvm-objdump.

For example, since all instructions are bundles, and bundles do not branch, branch targets could not be printed.

This patch modifies the Hexagon disassembler to return individual sub-instructions instead of entire bundles, enabling correct printing of branch targets and relocations. It also introduces `MCDisassembler::getInstructionBundle` for cases where the full bundle is still needed.

By default, llvm-objdump separates instructions with newlines. However, this does not work well for Hexagon syntax:

  { inst1
    inst2
    inst3
    inst4 <branch> } :endloop0

Instructions may be followed by a closing brace, a closing brace with `:endloop`, or a newline. Branches must appear within the braces.

To address this, `PrettyPrinter::getInstructionSeparator()` is added and overridden for Hexagon.

quic-areg committed:
[Hexagon][llvm-objdump] Improve disassembly of Hexagon bundles

Hexagon instructions are VLIW "bundles" of up to four instruction words
encoded as a single MCInst with operands for each sub-instruction.
Previously, the disassembler's getInstruction() returned the full bundle,
which made it difficult to work with llvm-objdump.

For example, since all instructions are bundles, and bundles do not branch,
branch targets could not be printed.

This patch modifies the Hexagon disassembler to return individual
sub-instructions instead of entire bundles, enabling correct printing
of branch targets and relocations. It also introduces
`MCDisassembler::getInstructionBundle` for cases where the full bundle is
still needed.

By default, llvm-objdump separates instructions with newlines. However,
this does not work well for Hexagon syntax:

  { inst1
    inst2
    inst3
    inst4 <branch> } :endloop0

Instructions may be followed by a closing brace, a closing brace with
`:endloop`, or a newline. Branches must appear within the braces.

To address this, `PrettyPrinter::getInstructionSeparator()` is added and
overridden for Hexagon.

jh7370 commented on llvm/lib/Target/Hexagon/MCTargetDesc/HexagonInstPrinter.cpp:
```diff
@@ -33,30 +33,17 @@ void HexagonInstPrinter::printRegName(raw_ostream &O, MCRegister Reg) {
+    OS << '\v';
+    HasExtender = false;
+    printInstruction(MI->getOperand(0).getInst(), Address, OS);
+  } else
```
Nit: llvm style is to put braces on all parts of an if/else, if at least one part requires it.

quic-areg replied:
Done

jh7370 commented on llvm/test/tools/llvm-objdump/ELF/Hexagon/branch-targets.yaml:
```diff
@@ -0,0 +1,26 @@
+## Check that branch targets are printed within instruction packets for Hexagon
```
```suggestion
## Check that branch targets are printed within instruction packets for Hexagon.
```

jh7370 commented on llvm/test/tools/llvm-objdump/ELF/Hexagon/branch-targets.yaml:
```diff
@@ -0,0 +1,26 @@
+# CHECK-NEXT:        8:       1e c0 1e 96     961ec01e {      dealloc_return }
+# CHECK-NEXT:        c:       00 c0 9d a0     a09dc000 {      allocframe(#0x0) }
+# CHECK-NEXT:       10:       00 c0 00 5a     5a00c000 {      call 0x10 <.text+0x10> }
+# CHECK-NEXT:       14:       1e c0 1e 96     961ec01e {      dealloc_return }
```
Nit: missing new line at EOF.

quic-areg replied:
Done

jh7370 reviewed (commented):
I don't know enough about Hexagon to be able to review this properly, but the suggestions sound reasonable.

Pre-merge CI is failing though.

quic-areg committed:
fix braces style

quic-areg committed:
add newline for test

MaskRay commented on llvm/tools/llvm-objdump/llvm-objdump.cpp:
```diff
@@ -693,6 +693,32 @@ class PrettyPrinter {
+                                       StringRef Comments,
+                                       LiveVariablePrinter &LVP) {
+    do {
+      if (!Comments.empty()) {
```
If Comments is a multi-line comment, `getInstructionSeparator();` will be printed multiple times, which is not correct. 

This is a case where we should probably just duplicate emitPostInstructionInfo for Hexagon and not introduce a new virtual function `getInstructionSeparator`

quic-areg committed:
- remove getInstructionSeparator
- test improvements:
  - multi-insn packets
  - branch targets
  - endloops
  - inline relocations

quic-areg committed:
fix empty check-next

androm3da commented on llvm/lib/Target/Hexagon/Disassembler/HexagonDisassembler.cpp:
```diff
@@ -171,43 +187,88 @@ LLVMInitializeHexagonDisassembler() {
+    Result = getSingleInstruction(*Inst, *CurrentBundle, Bytes, Address, CS,
+                                  Complete);
+    CurrentBundle->addOperand(MCOperand::createInst(Inst));
+    BytesToSkip += HEXAGON_INSTR_SIZE;
     Bytes = Bytes.slice(HEXAGON_INSTR_SIZE);
   }
   if (Result == MCDisassembler::Fail)
-    return Result;
-  if (Size > HEXAGON_MAX_PACKET_SIZE)
-    return MCDisassembler::Fail;
+    return false;
+  if (BytesToSkip > HEXAGON_MAX_PACKET_SIZE)
+    return false;
```
`BytesToSkip += HEXAGON_INSTR_SIZE;` - is this right?  Are you testing duplex instructions here too?  If so that might not be the right constant to use here.

`if (Bytes.size() < HEXAGON_INSTR_SIZE)` is probably safe though.

Aside: we should probably define independent `HEXAGON_MAX_PACKET_SIZE_BYTES` and `HEXAGON_MAX_PACKET_SIZE_INSTS` to clear this up.

quic-areg replied:
I think this is right. It is the same code we had before, moved from `getInstruction` to `makeBundle`. If `Inst` is a duplex of instructions, it should still take up a word like an ordinary instruction.

quic-seaswara commented:
I dont see any more review comments for this change, can this be merged ? 

@MaskRay / @androm3da / @jh7370 ?

quic-areg committed:
Merge branch 'main' into hexagon-objdump

jh7370 commented:
> I dont see any more review comments for this change, can this be merged ?
> 
> @MaskRay / @androm3da / @jh7370 ?

As noted earlier, I don't have enough Hexagon knowledge to feel comfortable approving myself.

MaskRay reviewed (commented):
Thanks for the delay... This looks great!

quic-seaswara commented:
@quic-areg  can you please cherry-pick this to the 21.0 release branch ?

quic-areg commented:
/cherry-pick ac7ceb3
